docs/proposals/2170-kubeflow-training-v2/README.md

```diff
@@ -0,0 +1,1432 @@
+    Containers []Container `json:"container,omitempty"`
+
+    // Override parameters for InitContainers.
+    InitContainer []Container `json:"initContainer,omitempty"`
+
+    // Override parameters for volumes.
+    Volumes []corev1.Volume `json:"volume,omitempty"`
+
+    // Custom Service Account
+    ServiceAccountName string `json:"serviceAccountName,omitempty"`
+}
+
+// Override for each container.
+// Parameters from Trainer, DatasetConfig, and ModelConfig will take precedence.
+type Container struct {
+
+    // Name for the container.
+    Name string `json:"name"`
+
+    // Command for the container.
+    Command []string `json:"command,omitempty" protobuf:"bytes,3,rep,name=command"`
+
+    // Args for the container.
+    Args []string `json:"args,omitempty"`
+
+    // Env for the container.
+    Env []corev1.EnvVar `json:"env,omitempty"`
+
+    // Env for the container.
+    EnvFrom []corev1.EnvFromSource `json:"envFrom,omitempty"`
+
+    // Override parameters for volume mounts.
+    VolumeMounts []VolumeMount `json:"volumeMounts,omitempty"`
+}
+```
+
+#### Example of TrainJob with Overrides
+
+This example shows how to override user-identity for sidecar container and add volume to the
+trainer container.
+
+```yaml
+apiVersion: kubeflow.org/v2alpha1
+kind: TrainJob
+metadata:
+  name: pytorch-distributed
+  namespace: tenant-alpha
+spec:
+  trainingRuntimeRef:
+    name: pytorch-distributed-gpu
+  trainer:
+    image: docker.io/custom-training
+  podSpecOverrides:
+    - targetReplicatedJobs:
+        - initializer
+          node
+      containers:
+        - name: user-identity
+          value: 123
+        - name: trainer
+          volumeMounts:
+            - name: user-123-volume
+              mountPath: /workspace
+      volumes:
+        - name: user-123-volume
+          persistentVolumeClaim:
+            claimName: user-123-volume
+```
+
+## The Training Runtime API
+
+The `TrainingRuntime` is the pre-created configurations of model training on the cluster,
+representing as blueprints. For example, Elastic PyTorch training, MPI DeepSpeed configuration,
+BERT LLM Fine-Tuning.
+
+These blueprints can be deployed within the Training Operator control plane and stored in a Kubeflow
+public repository that users can apply to their clusters.
+
+Platform or ML engineers can tweak existing blueprints, based on their requirements. For example,
+using custom configurations.
+
+The Kubeflow Training Operator can maintain more Training Runtimes when the community is ready to
+support them. For example, runtimes for [Jax](https://jax.readthedocs.io/en/latest/index.html) or
+[MLX](https://ml-explore.github.io/mlx/build/html/index.html). We will support PyTorch and MPI runtimes.
+After initial implementation, we will support TensorFlow, XGboost, and PaddlePaddle runtimes, but
+it is out of scope of this KEP.
+
+The `TrainingRuntime` is immutable, and so to make a change, a new version of the `TrainingRuntime`
```

Comment:
@tenzen-y How would immutability impact Kueue integration?

I'm trying to think through what kind of edits kueue or other controllers would want to add to TrainJob.

I know that for JobSet we had to relax quite a few immutability requirements for Kueue integration so I think its worth thinking through this.

Reply:
This immutable is affected only on the `TrainingRuntime` and `ClusterTrainingRuntime`. So, we can still modify the TrainJob. Hence, I think that this will not bring us any specific confusion in JobSet and Kueue.

@kannon92 If I am missing anything, please let me know.

Reply:
For example, in JobSet, we had to relax some of these rules:

https://github.com/kubernetes-sigs/jobset/blob/main/pkg/webhooks/jobset_webhook.go#L320

I just want to make sure that integration for Kueue has everything we need in TrainJob rather than TrainingRuntime. It sounds like Platform engineers define the TrainingRuntime and that template is reused for TrainJob.

But do we need to update anything in TrainingRuntime for Kueue?

Reply:
I imagined that Kueue modifies the PodOverrideSpec field: https://github.com/kubeflow/training-operator/blob/02947981c995e3aec0235bef28bd7fc2c6aa5a1a/docs/proposals/2170-kubeflow-training-v2/README.md#the-pod-spec-overrides-apis

But, it seems that the podSpecOverride doesn't have scheduling directive. We need to allow external controller to modify the scheduling directive so that Kueue can handle the objects.
Thanks for a great call!

@andreyvelich Could you add scheduling directives to the podSpecOverrides so that we can natively support Kueue?

- nodeSelectors
- tolerations
- schedulingGates

@kannon92 If I'm missing any scheduling directives, please let me know.

Reply:
Oh, additionally, the TrainJob does not expose to modify the managedBy and suspend field.

Reply:
Thank you for checking that out. I have no more concerns. LGTM.

Reply:
How will the TrainJob scheduling directives will work together with JobSet ? @tenzen-y Let's discuss it in more details tomorrow, so we can see how we should introduce those APIs.

Reply:
Sure.